pkg/lib/authn/authenticator/password/generator_test.go

```diff
@@ -0,0 +1,273 @@
+	Convey("Given a password generator with multiple requirements", t, func() {
+		generator := &Generator{
+			Checker:    &Checker{},
+			RandSource: &TestRandSource{},
+			Policy: &config.PasswordPolicyConfig{
+				MinLength:         newInt(12),
+				UppercaseRequired: true,
+				LowercaseRequired: true,
+				DigitRequired:     true,
+				SymbolRequired:    true,
+			},
+		}
+
+		password, err := generator.Generate()
+
+		Convey("should meet all requirements", func() {
+			So(checkPasswordUppercase(password), ShouldBeTrue)
+			So(checkPasswordLowercase(password), ShouldBeTrue)
+			So(checkPasswordDigit(password), ShouldBeTrue)
+			So(checkPasswordSymbol(password), ShouldBeTrue)
+			So(len(password), ShouldEqual, 12)
+			So(err, ShouldBeNil)
+		})
+	})
+}
+
+func TestMinLengthRequirement(t *testing.T) {
+	Convey("Given a password generator with a minimum length requirement", t, func() {
+		generator := &Generator{
+			Checker:    &Checker{},
+			RandSource: &TestRandSource{},
+			Policy: &config.PasswordPolicyConfig{
+				MinLength: newInt(40),
+			},
+		}
+
+		password, err := generator.Generate()
+
+		Convey("should meet the minimum length requirement", func() {
+			So(len(password), ShouldBeGreaterThanOrEqualTo, 40)
+			So(err, ShouldBeNil)
+		})
+	})
+}
+
+func TestMinGuessableLevelRequirement(t *testing.T) {
+	Convey("Given a password generator with a minimum guessable level requirement", t, func() {
+		generator := &Generator{
+			Checker:    &Checker{},
+			RandSource: &CryptoRandSource{},
+			Policy: &config.PasswordPolicyConfig{
+				MinLength:             newInt(8),
+				MinimumGuessableLevel: 4,
+			},
+		}
+
+		password, err := generator.Generate()
+
+		Convey("should meet the minimum guessable level requirement", func() {
+			level, _ := checkPasswordGuessableLevel(password, 4)
+			So(len(password), ShouldBeGreaterThanOrEqualTo, 32)
+			So(level, ShouldBeGreaterThanOrEqualTo, 4)
+			So(err, ShouldBeNil)
+		})
+	})
+}
+
+func TestExcludedKeywordsRequirement(t *testing.T) {
+	Convey("Given a password generator with excluded keywords", t, func() {
+		excluded := []string{"1", "2", "3"}
+		generator := &Generator{
+			Checker: &Checker{
+				PwExcludedKeywords: excluded,
+			},
+			RandSource: &CryptoRandSource{},
+			Policy: &config.PasswordPolicyConfig{
+				MinLength:        newInt(8),
+				DigitRequired:    true,
+				ExcludedKeywords: excluded,
+			},
+		}
+
+		password, err := generator.Generate()
+
+		Convey("should not contain any excluded keywords", func() {
+			So(checkPasswordExcludedKeywords(password, excluded), ShouldBeTrue)
+			So(err, ShouldBeNil)
+		})
+	})
+}
+
+func TestPrepareCharacterSet(t *testing.T) {
+	Convey("Given a password policy config", t, func() {
+		Convey("When no specific requirements are set", func() {
+			policy := &config.PasswordPolicyConfig{}
+			result, _ := prepareCharacterSet(policy)
+			So(result, ShouldContainSubstring, CharListLowercase)
+			So(result, ShouldContainSubstring, CharListDigit)
```

Comment:
Change the expectation to assert on the literal result. Not a substring. This style does not catch the bug in the algorithm (The bug of duplicate character set).